File: server/dukan/domain/catalog.py
"""Catalog domain. Mirrors packages/dukan_core/lib/domain/catalog.dart and
docs/domain/catalog.md. Simple products with multiple barcodes (no variants).
Quantities are integers in the unit's minor granularity (10^decimal_places).
"""

from __future__ import annotations

from dataclasses import dataclass

from dukan.domain.numbers import NumberProblem, parse_scaled
from dukan.shared.errors import ConflictError, ValidationError
from dukan.shared.limits import MONEY_MAX
from dukan.shared.money import Money


@dataclass(frozen=True, slots=True)
class Unit:
    id: str
    name: str
    decimal_places: int


@dataclass(frozen=True, slots=True)
class Category:
    id: str
    name: str
    parent_id: str | None = None


@dataclass(frozen=True, slots=True)
class Product:
    id: str
    sku: str
    name: str
    unit_id: str
    sell_price: Money
    category_id: str | None = None
    cost: Money | None = None
    track_stock: bool = True
    is_active: bool = True
    version: int = 1


@dataclass(frozen=True, slots=True)
class Barcode:
    id: str
    product_id: str
    code: str
    symbology: str = "ean13"


def assert_unique_sku(*, sku: str, taken: bool) -> None:
    if taken:
        raise ConflictError("PRODUCT_DUPLICATE_SKU", sku=sku)


def assert_unique_barcode(*, code: str, taken: bool) -> None:
    if taken:
        raise ConflictError("BARCODE_DUPLICATE", barcode=code)


def quantity_to_minor(value: str, decimal_places: int) -> int:
    """Parse a user quantity (Persian or Latin digits) into integer minor units.

    Raises ValidationError CATALOG_UNIT_PRECISION (too many decimals) or
    CATALOG_QTY_INVALID (not a number, or too large).
    """
    parsed, problem = parse_scaled(value, decimal_places)
    if problem is NumberProblem.TOO_PRECISE:
        raise ValidationError("CATALOG_UNIT_PRECISION", allowed=decimal_places)
    if parsed is None:
        raise ValidationError("CATALOG_QTY_INVALID", input=value[:32])
    return parsed


def assert_price_valid(*, sell_price_minor: int) -> None:
    """A selling price is zero or more: a free item is fine, a negative price
    would pay the customer. Raises ValidationError CATALOG_PRICE_INVALID."""
    if not 0 <= sell_price_minor <= MONEY_MAX:
        raise ValidationError("CATALOG_PRICE_INVALID", price=sell_price_minor)


def format_quantity(minor: int, decimal_places: int) -> str:
    if decimal_places == 0:
        return str(minor)
    negative = minor < 0
    magnitude = abs(minor)
    unit = 10**decimal_places
    whole = magnitude // unit
    frac = str(magnitude % unit).rjust(decimal_places, "0")
    return f"{'-' if negative else ''}{whole}.{frac}"

File: packages/dukan_core/lib/domain/catalog.dart
import '../shared/errors.dart';
import '../shared/money.dart';
import 'numbers.dart';

/// Catalog domain — simple products with multiple barcodes (no variants in
/// Phase 2). Mirrors docs/domain/catalog.md and server/dukan/domain/catalog.py.
/// Quantities are integers in the unit's minor granularity (10^decimalPlaces).

final class Unit {
  const Unit({required this.id, required this.name, required this.decimalPlaces});
  final String id;
  final String name;
  final int decimalPlaces; // 0 = whole units (piece), 3 = grams within a kg
}

final class Category {
  const Category({required this.id, required this.name, this.parentId});
  final String id;
  final String name;
  final String? parentId;
}

final class Product {
  const Product({
    required this.id,
    required this.sku,
    required this.name,
    required this.unitId,
    required this.sellPrice,
    this.categoryId,
    this.cost,
    this.trackStock = true,
    this.isActive = true,
    this.version = 0,
  });

  final String id;
  final String sku;
  final String name;
  final String unitId;
  final Money sellPrice;
  final String? categoryId;
  final Money? cost;
  final bool trackStock;
  final bool isActive;
  final int version;
}

final class Barcode {
  const Barcode({
    required this.id,
    required this.productId,
    required this.code,
    this.symbology = 'ean13',
  });
  final String id;
  final String productId;
  final String code;
  final String symbology;
}

/// Raises [ConflictError] `PRODUCT_DUPLICATE_SKU` when the sku is already used.
void assertUniqueSku({required String sku, required bool taken}) {
  if (taken) throw ConflictError('PRODUCT_DUPLICATE_SKU', {'sku': sku});
}

/// Raises [ConflictError] `BARCODE_DUPLICATE` when the barcode already resolves.
void assertUniqueBarcode({required String code, required bool taken}) {
  if (taken) throw ConflictError('BARCODE_DUPLICATE', {'barcode': code});
}

/// The codes a scanned barcode may be stored under, in the order to try them.
///
/// A UPC-A is the same product as the EAN-13 written with a zero in front, and
/// readers disagree about which they report: Android's ML Kit gives the 12
/// digits, Apple's Vision the 13, and a keyboard-wedge scanner whichever it was
/// set up to send. The code as read comes first, then its twin.
List<String> barcodeLookupCodes(String code) {
  if (!RegExp(r'^[0-9]+$').hasMatch(code)) return [code];
  if (code.length == 12) return [code, '0$code'];
  if (code.length == 13 && code.startsWith('0')) return [code, code.substring(1)];
  return [code];
}

int _pow10(int n) {
  var r = 1;
  for (var i = 0; i < n; i++) {
    r *= 10;
  }
  return r;
}

/// Parse a user-entered quantity (Persian or Latin digits) into integer minor
/// units for a unit with [decimalPlaces]. Raises [ValidationError]
/// `CATALOG_UNIT_PRECISION` when the input has more decimals than the unit
/// allows (e.g. 1.5 of a piece), or `CATALOG_QTY_INVALID` when it is not a
/// number (or too large).
int quantityToMinor(String input, int decimalPlaces) {
  final parsed = parseScaled(input, decimalPlaces);
  if (parsed.problem == NumberProblem.tooPrecise) {
    throw ValidationError('CATALOG_UNIT_PRECISION', {'allowed': decimalPlaces});
  }
  return parsed.value ?? (throw ValidationError('CATALOG_QTY_INVALID', {'input': input}));
}

/// A selling price is zero or more: a free item is fine, a negative price would
/// pay the customer. Raises [ValidationError] `CATALOG_PRICE_INVALID`.
void assertPriceValid({required int sellPriceMinor}) {
  if (sellPriceMinor < 0 || sellPriceMinor > moneyMax) {
    throw ValidationError('CATALOG_PRICE_INVALID', {'price': sellPriceMinor});
  }
}

/// Render integer minor units as a decimal string for a unit's [decimalPlaces].
String formatQuantity(int minor, int decimalPlaces) {
  if (decimalPlaces == 0) return minor.toString();
  final negative = minor < 0;
  final magnitude = minor.abs();
  final unit = _pow10(decimalPlaces);
  final whole = magnitude ~/ unit;
  final frac = (magnitude % unit).toString().padLeft(decimalPlaces, '0');
  return '${negative ? '-' : ''}$whole.$frac';
}

/// The units every shop starts with. Their ids are the same on every device and
/// on the server (server/dukan/domain/catalog.py BUILTIN_UNITS), so seeding them
/// again never makes a second "kg".
const builtInUnits = <Unit>[
  Unit(id: '00000000-0000-7000-8000-000000000001', name: 'piece', decimalPlaces: 0),
  Unit(id: '00000000-0000-7000-8000-000000000002', name: 'kg', decimalPlaces: 3),
  Unit(id: '00000000-0000-7000-8000-000000000003', name: 'litre', decimalPlaces: 3),
  Unit(id: '00000000-0000-7000-8000-000000000004', name: 'dozen', decimalPlaces: 0),
  Unit(id: '00000000-0000-7000-8000-000000000005', name: 'meter', decimalPlaces: 2),
];
